### PR TITLE
Make threadDelay duration unbounded

### PR DESCRIPTION
**Pull request by kderme**
https://github.com/input-output-hk/ouroboros-network/issues/2026

Timeout duration is still bounded by Int

A ci failure was detected https://hydra.iohk.io/build/2668383/nixlog/1. It's related to threadDelay, but I'm not sure if it's related to this pr.

### CODE REVIEW

File: io-sim-classes/src/Control/Monad/Class/MonadTimer.hs [hunk 1]
```diff
@@ -178,16 +178,26 @@ instance MonadTimer IO where
...
 
 diffTimeToMicrosecondsAsInt :: DiffTime -> Int
 diffTimeToMicrosecondsAsInt d =
     let usec :: Integer
-        usec = diffTimeToPicoseconds d `div` 1000000 in
+        usec = diffTimeToMicrosecondsAsInteger d in
     -- Can only represent usec times that fit within an Int, which on 32bit
     -- systems means 2^31 usec, which is only ~35 minutes.
     assert (usec <= fromIntegral (maxBound :: Int)) $
     fromIntegral usec
```
Comment (mrBliss) [inline io-sim-classes/src/Control/Monad/Class/MonadTimer.hs:195]: This function is the cause of the problem: the assert can fail.

You address it for `threadDelay`, but the problematic function is used by other functions as well.

I don't think @dcoutts (who wrote `io-sim(-classes)`) would be satisfied that we just fix one function as opposed to fixing it properly once and for all.

However, I think we, consensus, can work around the issue entirely on our side.

Plan:
* Undo the changes to `MonadTimer`
* Add a function to consensus:
  ```haskell
  -- | Document that this is for sleeping longer periods, mention the overflow and ~35 minute limit
  threadSleep :: MonadTimer m => DiffTime -> m ()
  threadSleep = <do the loop>
  ```
  (I first had a new `MonadSleep` class in mind, but a function is actually enough).
* Look for all uses of `threadDelay` in consensus and replace them with `threadSleep` where the chance exists that we might have to wait longer than 35 minutes. In particular, the place where we wait for the blockchain time to start.

Comment (kderme) [reply io-sim-classes/src/Control/Monad/Class/MonadTimer.hs:195]: Wouldn't it better to leave the changes in MonadTImer? This way all users of the class will have the fix, without needing any wrapper.

This function is only used on Timeout, other than threadDelay. I could take a look if this is also fixable, if we want to entirely remove it.

Comment (mrBliss) [reply io-sim-classes/src/Control/Monad/Class/MonadTimer.hs:195]: > Wouldn't it better to leave the changes in MonadTImer? This way all users of the class will have the fix, without needing any wrapper.

Sure, that would be ideal. But I'm not sure how easy that is.

> This function is only used on Timeout, other than threadDelay. I could take a look if this is also fixable, if we want to entirely remove it.

In that case you'll have to ask a review from @dcoutts.


First do the `threadSleep` fix, which should be merged ASAP, as people launching testnets that start in more than 35 minutes will run into the bug.

Comment (kderme) [reply io-sim-classes/src/Control/Monad/Class/MonadTimer.hs]: I replaced threadDelay with threadSleep in some places, hopefully I didn't miss something important.

Also I assumed your quotes and replies are reversed at your previous response?

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/IOLike.hs [hunk 2]
```diff
@@ -91,3 +93,17 @@ class ( MonadAsync              m
...
+
+-- | This can be used to sleep for longer periods of time. 'threadDelay' can't
+-- be used for delays over ~35mins, because it uses an 'Int', which can
+-- overflow. See 'diffTimeToMicrosecondsAsInt' for more info.
+threadSleep :: MonadDelay m => DiffTime -> m ()
+threadSleep time = do
+    threadDelay maxDelay
+    when (maxDelay /= time) $ threadSleep (time - maxDelay)
+  where
+    maxDelay = min time $ fromIntegral (maxBound :: Int)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/IOLike.hs:109]: I find it confusing that `maxDelay` is not really the maximum delay that we can wait for, both the minium of the maximum delay we can wait for and the requested wait.

What about:
```suggestion
threadSleep time
    | time > maxDelay
    = threadDelay maxDelay *> threadSleep (time - maxDelay)
    | otherwise
    = threadDelay time
  where
    maxDelay :: DiffTime
    maxDelay = fromIntegral (maxBound :: Int)
```

File: network-mux/src/Network/Mux/Timeout.hs [hunk 3]
```diff
@@ -94,7 +94,7 @@ withTimeoutSerialNative body = body MonadTimer.timeout
 --
 -- The implementation here only relies upon 'threadDelay' which has an
 -- efficient implementation on all platforms. It uses a separate monitoring
--- thread which will through an exception to terminate the action if the
+-- thread which will throw an exception to terminate the action if the
```
Comment (mrBliss) [inline network-mux/src/Network/Mux/Timeout.hs:97]: This is unrelated to the PR, create a separate commit (a separate PR is not needed) for it.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/IOLike.hs [hunk 4]
```diff
@@ -34,6 +34,8 @@ module Ouroboros.Consensus.Util.IOLike (
   , diffTime
     -- *** MonadDelay
   , MonadDelay(..)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/IOLike.hs:36]:
```suggestion
  , MonadDelay
```
Idea: just export `MonadDelay` but not its sole method `threadDelay`). This will prevent us from accidentally using `threadDelay` in the future.

It's still not perfect, as in some places we might import the following directly:
```haskell
import           Control.Monad.Class.MonadTimer (MonadDelay (..))
```

File: ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime/WallClock/Util.hs [hunk 5]
```diff
@@ -21,6 +21,7 @@ import           Control.Monad.Class.MonadTimer (MonadDelay (..))
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime/WallClock/Util.hs:21]: Remove this import (and import `MonadDelay` from `IOLIke`, because `threadDelay` is now still in scope).

File: ouroboros-consensus/src/Ouroboros/Consensus/Util/IOLike.hs [hunk 6]
```diff
@@ -91,3 +93,20 @@ class ( MonadAsync              m
...
+
+{-------------------------------------------------------------------------------
+  MonadDelay Wrappers
+-------------------------------------------------------------------------------}
+
+-- | This can be used to sleep for longer periods of time. In some cases,
+-- 'threadDelay' can't be used for delays over ~35mins, because it uses an
+-- 'Int', which can overflow. See 'diffTimeToMicrosecondsAsInt' for more info.
+threadSleep :: MonadDelay m => DiffTime -> m ()
+threadSleep time
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util/IOLike.hs:105]: An even better idea from @edsko: call this one `threadDelay` so that all other changes in this PR are no longer needed and that we reduce the chances of accidentally importing/using the broken `threadDelay` (typically you import `IOLike` too, so you'd have a conflict).

File: ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/Simple.hs [hunk 7]
```diff
@@ -13,6 +13,7 @@
 module Test.Consensus.BlockchainTime.Simple (tests) where
 
 import           Control.Exception (SomeException, fromException)
+import           Control.Monad.Class.MonadTimer (MonadDelay (..))
```
Comment (mrBliss) [inline ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/Simple.hs:16]: Remove this import

Comment (kderme) [reply ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/Simple.hs:16]: I think this is needed for the instance declaration.

Comment (mrBliss) [reply ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/Simple.hs:16]: What error do you get when you remove it?

Comment (kderme) [reply ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/Simple.hs]:
```
test-consensus/Test/Consensus/BlockchainTime/Simple.hs:324:3: error:
    ‘threadDelay’ is not a (visible) method of class ‘MonadDelay’
    |
324 |   threadDelay d = OverrideDelay $ ReaderT $ \schedule -> do
    |   ^^^^^^^^^^^

test-consensus/Test/Consensus/BlockchainTime/Simple.hs:346:3: error:
    ‘threadDelay’ is not a (visible) method of class ‘MonadDelay’
    |
346 |   threadDelay d = OverrideDelay $ ReaderT $ \_schedule -> threadDelay d
    |   ^^^^^^^^^^^

```

If I import it qualified I get scope to the method, even though I never actually use the module qualified (qualified names are now allowed in bindings positions).

File: ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/Simple.hs [hunk 8]
```diff
@@ -13,6 +13,7 @@
 module Test.Consensus.BlockchainTime.Simple (tests) where
 
 import           Control.Exception (SomeException, fromException)
+import qualified Control.Monad.Class.MonadTimer
```
Comment (kderme) [inline ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/Simple.hs:16]:
```suggestion
import qualified Control.Monad.Class.MonadTimer as Unused
```
Is this better?

Comment (mrBliss) [reply ouroboros-consensus/test-consensus/Test/Consensus/BlockchainTime/Simple.hs:16]: I see. Ok.

But move it down so it's together with
```haskell
import           Control.Monad.Class.MonadTime
import           Control.Monad.IOSim
```

**Comment by mrBliss** [comment]
bors merge
